### PR TITLE
[GEOS-9646]: INSPIRE validation get errors of GetMapRequest parameters

### PR DESCRIPTION
**Pull request by ahuarte47**
JIRA issue:
https://osgeo-org.atlassian.net/browse/GEOS-9646

The [INSPIRE validator](https://inspire.ec.europa.eu/validator) throws some errors testing GeoServer WMS Services.

The GetMap validation report shows three errors, all of them about input parameters:

image

We get these error messages:
- A GetMap was requested without mandatory parameter 'STYLES' and the service did not throw an exception.
- A GetMap was requested without mandatory parameter 'CRS' and the service did not throw an exception.
- A GetMap was requested using parameter 'TRANSPARENT' with an invalid value and the service did not throw an exception.

This code fixes all of them taking care of CITE compliant mode.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

**Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.**

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for community modules):
- [x] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [x] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [x] New unit tests have been added covering the changes
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by travis-ci after opening this PR)
- [ ] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates (screenshots, text)
- [ ] Commits changing the REST API, or any configuration object, should check if the REST API docs (Swagger YAML files and classic documentation) need to be updated.

### CODE REVIEW

**Comment by aaime** [comment]
The code changes seem good. Too bad you did not read the template for the PR and removed it instead:

```
<Include a few sentences describing the overall goals for this Pull Request. Please do not remove the checklist below, pull requests missing it will be ignored.>
```

So please re-instate [the template](https://github.com/geoserver/geoserver/blob/master/.github/PULL_REQUEST_TEMPLATE.md) and satisfy the checklist, or this PR will indeed be ignored :-D

**Comment by ahuarte47** [comment]
Sorry @aaime, I was waiting travis tests to finally fill every checkboxes, I have already include it.

**Comment by aaime** [comment]
I see you're missing a Jira ticket, and the Junit tests.

**Comment by ahuarte47** [comment]
Thanks @aaime, I have already added them

File: src/wms/src/main/java/org/geoserver/wms/map/GetMapKvpRequestReader.java [hunk 1]
```diff
@@ -243,12 +243,27 @@ public GetMapRequest read(Object request, Map kvp, Map rawKvp) throws Exception
         // set the raw params used to create the request
         getMap.setRawKvp(rawKvp);
 
+        boolean citeCompliant = wms.getServiceInfo().isCiteCompliant();
+
         // wms 1.3, srs changed to crs
         if (kvp.containsKey("crs")) {
             getMap.setSRS((String) kvp.get("crs"));
+        } else if (citeCompliant) {
```
Comment (aaime) [inline src/wms/src/main/java/org/geoserver/wms/map/GetMapKvpRequestReader.java:251]: Need to check the version of the protocol, this exception cannot be thrown if the version is any 1.x

File: src/wms/src/main/java/org/geoserver/wms/map/GetMapKvpRequestReader.java [hunk 2]
```diff
@@ -288,13 +303,19 @@ public GetMapRequest read(Object request, Map kvp, Map rawKvp) throws Exception
         if (layerParam != null) {
             List<String> layerNames = KvpUtils.readFlat(layerParam);
             requestedLayerInfos.addAll(parseLayers(layerNames, remoteOwsUrl, remoteOwsType));
+        } else if (citeCompliant) {
```
Comment (aaime) [inline src/wms/src/main/java/org/geoserver/wms/map/GetMapKvpRequestReader.java:306]: The SLD extensions to WMS allow a request not to have layers, as long as a full SLD is specified either using &sld or &sld_body. The error must not be thrown in these conditions (which are probably not what INSPIRE had in mind, but nonetheless a OGC specification

File: src/wms/src/main/java/org/geoserver/wms/map/GetMapKvpRequestReader.java [hunk 3]
```diff
@@ -288,13 +303,19 @@ public GetMapRequest read(Object request, Map kvp, Map rawKvp) throws Exception
...
+            throw new ServiceException(
+                    "GetMap LAYERS parameter is missing, it is mandatory for CITE compliant mode");
         }
 
         // raw styles parameter
         String stylesParam = (String) kvp.get("STYLES");
         List<String> styleNameList = new ArrayList<String>();
         if (stylesParam != null) {
             styleNameList.addAll(KvpUtils.readFlat(stylesParam));
+        } else if (citeCompliant) {
```
Comment (aaime) [inline src/wms/src/main/java/org/geoserver/wms/map/GetMapKvpRequestReader.java:316]: Same as above for LAYERS, styles becomes non mandatory if SLD or SLD_BODY are specified.

Comment (ahuarte47) [reply src/wms/src/main/java/org/geoserver/wms/map/GetMapKvpRequestReader.java]: Thank you very much @aaime, I have changed the code and I have added more tests.

File: src/wms/src/main/java/org/geoserver/wms/map/GetMapKvpRequestReader.java (hunk 1 above)
Comment (ahuarte47) [reply src/wms/src/main/java/org/geoserver/wms/map/GetMapKvpRequestReader.java]: Done, thanks @aaime

**Comment by ahuarte47** [comment]
When reviewers accept this PR, I would like to propose the [backport to 2.17.x branch](https://github.com/geoserver/geoserver/pull/4317).

File: src/wms/src/main/java/org/geoserver/wms/map/GetMapKvpRequestReader.java [hunk 4]
```diff
@@ -243,12 +243,30 @@ public GetMapRequest read(Object request, Map kvp, Map rawKvp) throws Exception
...
+                    "GetMap CRS parameter is missing, it is mandatory for CITE compliant mode");
         }
         // do some additional checks
 
+        if (citeCompliant && rawKvp != null && rawKvp.containsKey("transparent")) {
+            String trans = (String) rawKvp.get("transparent");
+
+            if (!trans.equalsIgnoreCase("false")
+                    && !trans.equalsIgnoreCase("true")
+                    && !trans.equalsIgnoreCase("0")
```
Comment (aaime) [inline src/wms/src/main/java/org/geoserver/wms/map/GetMapKvpRequestReader.java:262]: See above, "transparent" does not seem to allow for "0" or "1", where are they coming from?

Quoting from the specification:

> The optional TRANSPARENT parameter specifies whether the map background is to be made transparent or not.
> TRANSPARENT can take on two values, "TRUE" or "FALSE". The default value is FALSE if this parameter is
> absent from the request.

Comment (ahuarte47) [reply src/wms/src/main/java/org/geoserver/wms/map/GetMapKvpRequestReader.java]: Thank you very much @aaime for your review and support.

**Comment by ahuarte47** [comment]
@aaime, thank you very much for your support & help!

**Comment by ahuarte47** [comment]
Hi @aaime, this fix raises a problem in the OpenLayer link in the PreviewLayers Page. Now GeoServer throws error (We get the " GetMap STYLES parameter is mandatory if SLD nor SLD_BODY are not specified" message) because the URL of the GetMap request in these links do not contain the "STYLES" parameter.

**Comment by ahuarte47** [comment]
I do not know if [here](https://github.com/geoserver/geoserver/blob/master/src/web/demo/src/main/java/org/geoserver/web/demo/PreviewLayer.java#L227) is the best place where appending the missing "STYLES" parameter (Even CITE compliance is not activated).

**Comment by ahuarte47** [comment]
[This ](https://github.com/ahuarte47/geoserver/commit/ae9cace5889e3ac735cc62d81ff25a9cea57ce32) could fix the issue, @aaime, dou you agree? Can I create a new PR with it?

**Comment by ahuarte47** [comment]
> Hi @aaime, this fix solves a problem in the OpenLayer link in the LayerPreview Page. Now GeoServer throws error (We get the " GetMap STYLES parameter is mandatory if SLD nor SLD_BODY are not specified" message) because the URL of the GetMap request in these links do not contain the "STYLES" parameter.

@aaime, Sorry, I'm worried if this issue can throw problems, what I have to do to try fix it? Do I have to create a new issue in JIRA?

**Comment by aaime** [comment]
You can change the preview URLs to contain an empty styles parameter.

**Comment by ahuarte47** [comment]
Do you agree I add a new PR with [this change](https://github.com/ahuarte47/geoserver/commit/ae9cace5889e3ac735cc62d81ff25a9cea57ce32)? Sorry I misunderstood you, or if you prefer you can revert this PR in master and 2.17.x branches.

**Comment by aaime** [comment]
Go ahead with the new PR.

**Pull request merged**